app/models/metrics/base_metric.rb

```diff
@@ -0,0 +1,116 @@
+require 'csv'
+require 'json'
+
+module Metrics
+  module BaseMetric
+    attr_reader :label, :file, :interval, :timestamp,
+                :columns_name, :columns_field, :columns,
+                :rows_name, :rows_field, :rows,
+                :values
+
+    FORMATS = {
+      to_csv: 'data.csv',
+      to_fixed_key_json: 'data-fixed_key.json',
+      to_datasette_json: 'data-datasette.json',
+      to_d3_json: 'data-d3.json',
+    }.freeze
+
+    def setup_metric(metric)
+      @label = metric[:label]
+      @file = metric[:file]
+      @interval = metric[:interval]
+      @timestamp = nil
+      @columns_name = metric[:columns][:name]
+      @columns_field = metric[:columns][:field]
+      @columns = evaluate(metric[:columns][:values])
+      @rows_name = metric[:rows][:name]
+      @rows_field = metric[:rows][:field]
+      @rows = evaluate(metric[:rows][:values])
+      @values = nil
+    end
+
+    def to_csv
+      calculate_all_values
+      CSV.generate do |csv|
+        csv << [label] + columns.map { |column| column_key(column) }
+        rows.each do |row|
+          csv << [row_key(row)] + columns.map { |column| value(column, row) }
+        end
+      end
+    end
+
+    def to_fixed_key_json
+      calculate_all_values
+      JSON.pretty_generate({
+        label: label,
+        timestamp: timestamp.iso8601,
+        data: rows.map do |row|
+          {
+            row: row_key(row),
+            values: columns.map { |column| { column: column_key(column), value: value(column, row) } },
+          }
+        end,
+      })
+    end
+
+    def to_datasette_json
+      calculate_all_values
+      JSON.pretty_generate({
+        database: label,
+        timestamp: timestamp.iso8601,
+        columns: columns.map { |column| column_key(column) },
+        rows: rows.map { |row| columns.map { |column| value(column, row) } },
+      })
+    end
+
+    def to_d3_json
+      calculate_all_values
+      JSON.pretty_generate({
+        label: label,
+        timestamp: timestamp.iso8601,
+        data: rows.map { |row| ([[rows_name, row_key(row)]] + columns.map { |column| [column_key(column), value(column, row)] }).to_h },
+      })
+    end
+
+  private
+
+    def evaluate(var)
+      if var.is_a?(Proc)
+        var.call
+      else
+        var
+      end
+    end
+
+    def calculate_all_values
+      if values.nil? || timestamp.nil? || timestamp + interval < Time.zone.now
+        @values = {}.tap do |v|
+          columns.each do |column|
+            rows.each do |row|
+              v[value_key(column, row)] = calculate(column, row)
```

Comment:
I can envision a bit of data inconsistency here if I understood the flow correctly: we will calculate each query according to the required row/column values. Due to the time it will take for each query to compute, we might get different data due to the time elapsed between each row. Maybe there is something here around getting the data first frozen at that required time, then parsing the data in memory instead?

Reply:
Happy to discuss/confirm over a call might be easier

Reply:
Hmmm... that is a very interesting point and something I didn't consider. Lets have a chat about it.

Reply:
I've had a look through/tested most of the queries required for the metrics and there was a recurring theme: we want to count the number of moves grouped by a column. 

For instance testing one of the metric generators for `CountsByMoveType`, we will have to run the following queries:
```
[5] pry(main)> Metrics::Moves::CountsByMoveType.new.to_datasette_json
   (127.4ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "court_appearance"]]
   (108.7ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "court_other"]]
   (144.3ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "hospital"]]
   (113.2ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "police_transfer"]]
   (109.8ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "prison_recall"]]
   (108.5ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "prison_remand"]]
   (124.4ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "prison_transfer"]]
   (111.3ms)  SELECT COUNT(*) FROM "moves" WHERE "moves"."move_type" = $1  [["move_type", "video_remand"]]
```
One for each column of types. Instead we can run the following single query which would be much more performant and retrieves all the information we require:
```
[8] pry(main)> Move.group(:move_type).count
   (139.2ms)  SELECT COUNT(*) AS count_all, "moves"."move_type" AS moves_move_type FROM "moves" GROUP BY "moves"."move_type"
=> {"court_appearance"=>139770,
 "court_other"=>371,
 "hospital"=>25,
 "police_transfer"=>1238,
 "prison_recall"=>24731,
 "prison_remand"=>12697,
 "prison_transfer"=>26010,
 "video_remand"=>6922}
```
I suggest remodelling the metric generator to first fetch this data, and then manipulate this structure to accommodate the required data representation, rather than perform a query on each column/row type. 
The following queries can be done on the other metric generators example:
```
Move.group(:move_type).count
Move.group(:move_type, :status).count
Move.group(:status).count
Move.group(:supplier).count
Move.group(:supplier, :status).count
Move.group(:supplier, :move_type).count
```
Regarding the timeframe queries, something similar which includes the time range query and the grouping can be used example:
```
Move.where('date >= ?', '2020-12-09').group(:status).count
```
Which would cut down those queries drastically as there are quite a few there. Happy to discuss on a call. Also having a simpler query can help DRY up some of the code as well.